clash-lib/src/Clash/Core/Evaluator/Models.hs

```diff
@@ -0,0 +1,258 @@
+insertEnv scope i etv env =
+  case scope of
+    LocalId -> env
+      { envLocals = Map.insert i' etv (envLocals env)
+      , envInScope = extendInScopeSet (envInScope env) i'
+      }
+
+    GlobalId ->
+      let b = Binding i' noSrcSpan Inline etv
+       in env
+            { envGlobals = extendVarEnv i' b (envGlobals env)
+            , envInScope = extendInScopeSet (envInScope env) i'
+            }
+ where
+  i' = uniqAway (envInScope env) i
+
+insertAllEnv :: IdScope -> [(Id, Either Term Value)] -> Env -> Env
+insertAllEnv s xs env = foldr (uncurry $ insertEnv s) env xs
+
+-- | Add a new type to the environment. The unique for the new
+-- element is guaranteed to be unique within the environment.
+--
+insertEnvTy :: TyVar -> Type -> Env -> Env
+insertEnvTy i ty env = env
+  { envTypes = Map.insert i' ty (envTypes env)
+  , envInScope = extendInScopeSet (envInScope env) i'
+  }
+ where
+  i' = uniqAway (envInScope env) i
+
+insertAllEnvTys :: [(TyVar, Type)] -> Env -> Env
+insertAllEnvTys xs env = foldr (uncurry insertEnvTy) env xs
+
+-- | Add a new primitive to the environment. Primitives are keyed by an
+-- integer ID in the evaluator. If the prim already exists in the environment,
+-- you should call 'updateEnvPrim' instead.
+--
+insertEnvPrim :: Value -> Env -> Env
+insertEnvPrim v env =
+  env { envPrimsIO = (IntMap.insert n v pm, n + 1) }
+ where
+  (pm, n) = envPrimsIO env
+
+-- | Delete the element with the speicified Id from the environment.
+--
+deleteEnv :: IdScope -> Id -> Env -> Env
+deleteEnv scope i env =
+  case scope of
+    LocalId -> env { envLocals = Map.delete i (envLocals env) }
+    GlobalId -> env { envGlobals = delVarEnv (envGlobals env) i }
+
+updateEnvPrim :: Int -> Value -> Env -> Env
+updateEnvPrim i v env =
+  env { envPrimsIO = (IntMap.insert i v pm, n) }
+ where
+  (pm, n) = envPrimsIO env
+
+-- | Neutral terms cannot be reduced, as they represent things like variables
+-- which are unknown, partially applied functions, or case expressions where
+-- the scrutinee is not yet an inspectable value. Consider:
+--
+-- v              Stuck if "v" is a free variable
+-- p x1 .. xn     Stuck if "p" has no primitive reduction rule
+-- x $ y          Stuck if "x" is not known to be a lambda
+-- x @ A          Stuck if "x" is not known to be a type lambda
+-- case x of ...  Stuck if "x" is neutral (cannot choose an alternative)
+--
+data Neutral a
+  = NeVar   (Var Term)
+  | NePrim  PrimInfo [Either Value Type]
+  | NeApp   (Neutral a) a
+  | NeTyApp (Neutral a) Type
+  | NeCase  a Type [(Pat, a)]
+  deriving (Show, Generic, NFData)
+
+-- A term which has been normalised to weak head normal form (WHNF). This has
+-- no redexes at the head of the term, but subterms may still contain redexes.
+--
+data Value
+  = VNeu    (Neutral Value)
+  | VData   DataCon [Either Value Type]
+  | VLit    Literal
+  | VPrim   PrimInfo [Either Value Type]
+  | VLam    Id Term Env
```

Comment:
Why do only lambdas have an environment? why not all (with the exception of Literals)?

Reply:
Added `TODO` comment for explaining this when the semantics are in (for cross-referencing)